### PR TITLE
Optimize GatherElements further, add threshold for parallelizing Scaler.

### PR DESCRIPTION
**Pull request by pranavsharma**
**Description**: Optimize GatherElements further by parallelizing inner loop, add threshold for parallelizing Scaler.

**Motivation and Context**
Reduces runtime by 50% when indices exceed 500K. Follow up to [this PR](https://github.com/microsoft/onnxruntime/pull/5543).

### CODE REVIEW

File: onnxruntime/core/providers/cpu/tensor/gather_elements.cc [hunk 1]
```diff
@@ -135,23 +136,37 @@ static void core_impl(const Tensor* input_tensor, const Tensor* indices_tensor,
...
-  Tin indices_counter = -1;
-  int64_t output_counter = -1;
+  Tin indices_counter = 0;
   size_t element_size = input_tensor->DataType()->Size();
 
   std::vector<int64_t> process_dims(input_rank, 0);
+  int64_t output_counter = 0;
+
+  auto conditional_batch_call = [ttp, inner_dim_size](std::function<void(ptrdiff_t)> f) {
+    if (inner_dim_size < 10 * 1000) {  // TODO: tune this, arbitrary threshold
```
Comment (hariharans29) [inline onnxruntime/core/providers/cpu/tensor/gather_elements.cc:162]: Do we need a simple test for both the ops that invokes the parallel conditional branch (i.e.) >= 10k elements ?

File: onnxruntime/core/providers/cpu/tensor/gather_elements.cc [hunk 2]
```diff
@@ -135,23 +136,37 @@ static void core_impl(const Tensor* input_tensor, const Tensor* indices_tensor,
...
+      for (int64_t i = 0; i < inner_dim_size; ++i) {
+        f(i);
+      }
+    } else {
+      concurrency::ThreadPool::TryBatchParallelFor(ttp, inner_dim_size, f, 0);
```
Comment (hariharans29) [inline onnxruntime/core/providers/cpu/tensor/gather_elements.cc:167]: Do we need a check to see if ttp is null (i.e.) user forces single-threaded execution ?

Comment (pranavsharma) [reply onnxruntime/core/providers/cpu/tensor/gather_elements.cc:167]: TryBatchParalleFor does check for ttp to be nullptr. See https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/platform/threadpool.h#L169.

File: onnxruntime/core/providers/cpu/tensor/gather_elements.cc (hunk 1 above)
Comment (pranavsharma) [reply onnxruntime/core/providers/cpu/tensor/gather_elements.cc:162]: sure, will do.

File: onnxruntime/core/providers/cpu/ml/scaler.cc [hunk 3]
```diff
@@ -85,7 +87,7 @@ common::Status ScalerOp<T>::Compute(OpKernelContext* context) const {
   int64_t stride = x_dims.size() == 1 ? x_dims[0] : x_dims[1];
   auto* ttp = context->GetOperatorThreadPool();
   auto conditional_batch_call = [ttp, x_size](std::function<void(ptrdiff_t)> f) {
-    if (x_size < 10 * 1000) {  // TODO: tune this, arbitrary threshold
+    if (x_size < kParallelizationThreshold) {  // TODO: tune this, arbitrary threshold
```
Comment (xadupre) [inline onnxruntime/core/providers/cpu/ml/scaler.cc:90]: I wonder if this decision could not be handled in ThreadPool. TryBatchParallelFor can handle a more complex cost.

File: onnxruntime/core/providers/cpu/tensor/gather_elements.cc [hunk 4]
```diff
@@ -159,7 +161,7 @@ static void core_impl(const Tensor* input_tensor, const Tensor* indices_tensor,
   int64_t output_counter = 0;
 
   auto conditional_batch_call = [ttp, inner_dim_size](std::function<void(ptrdiff_t)> f) {
-    if (inner_dim_size < 10 * 1000) {  // TODO: tune this, arbitrary threshold
+    if (inner_dim_size < kParallelizationThreshold) {  // TODO: tune this, arbitrary threshold
```
Comment (xadupre) [inline onnxruntime/core/providers/cpu/tensor/gather_elements.cc:164]: I saw below this parallelization may be called multiple times from a loop. I was wondering if there was a way to call TryBatchParallelFor only once.

Comment (pranavsharma) [reply onnxruntime/core/providers/cpu/tensor/gather_elements.cc:164]: In the model I was working on the inner loop is the biggest contributor to runtime. In fact the outer loop runs only once. Hence, I didn't rework this more than what was required.

File: onnxruntime/core/providers/cpu/ml/scaler.cc (hunk 3 above)
Comment (pranavsharma) [reply onnxruntime/core/providers/cpu/ml/scaler.cc:90]: Yes, once we learn more we can convert this into a proper cost based model. For now, just relying on a crude heuristic.

**Review by hariharans29 — APPROVED** [review]

**Pull request merged**